### PR TITLE
INSP: remove some false positives from RsSelfConventionInspection

### PR DESCRIPTION
**Pull request by Kobzol**
The last PR that changed `RsSelfConventionInspection` (https://github.com/intellij-rust/intellij-rust/pull/6842) introduced some false positives. Based on discussion in https://github.com/intellij-rust/intellij-rust/issues/7475, in this PR I tried to align the inspection better with [Clippy](https://rust-lang.github.io/rust-clippy/master/index.html#wrong_self_convention).

What has changed:
- `with` convention now allows functions without any `self` (`Vec::with_capacity` etc.)
- `to` convention now allows `self` or `&self`
- `to_*_mut` convention now allows `&mut self`
- `get_*_mut` convention now allows `&mut self`

Fixes: https://github.com/intellij-rust/intellij-rust/issues/7475
Fixes: https://github.com/intellij-rust/intellij-rust/issues/7520

changelog: Reduce amount of false positives caused by `RsSelfConventionInspection`.

### CODE REVIEW

**Comment by afetisov** [comment]
I don't think there should be any conventions for `with_` methods. As a random data point, I have search the library dependencies in my project for function definitions starting with `with_` and various self conventions. For `&self` and `&mut self` my specific regexes were

```regex
fn\s+with\_[^(]+\(\s*&self
fn\s+with\_[^(]+\(\s*&mut self
```

I have also tried searching with extra lifetimes or whitespace, which had no effect on my specific dataset. The specific libraries are somewhat arbitrary, but include many popular big libraries such as `tokio` or `actix-web` and similarly popular transitive dependencies. We could repeat the search if more representative sample is available.

As expected, there were many methods with both `self` and no self at all. Both were related to the builders. Methods with no self appeared in many form, but a notable case were builders with a factory function which had the form `fn with_??<F>(f: F) where F: Factory`.

For some numbers, I have found 127 functions with `self` convention, and 422 with no receiver (although I don't guarantee that all of them are static methods and not free functions).

With respect to the current PR more interesting are the `&self ` and `&mut self` methods. Here I have found the following cases.

* `&self` --- 79 results. 
    * chrono --- 60 results. Methods like `fn with_month(&self, month: u32) -> Option<NaiveDate>` with months, days, hours etc in `impl NaiveDate`. Essentially those are methods of a builder pattern, but since `NaiveDate: Copy` it doesn't matter whether those methods take `self` or `&self`.
    * crossbeam-epoch --- 2 results. Methods `UnsafeCell::with_mut` and `Shared::with_tag`. The former is a builder method on a type with interior mutability, so it takes a `&self`. The latter is implemented  on a `Shared: Copy`, so again `self` vs `&self` is an arbitrary design decision.
    * net2 --- 2 results. Methods `TcpBuilder::with_socket` and `UdpBuilder::with_socket`. Both are builder methods on a type with interior mutability.

    * std --- 4 results. That's an interesting one. 

        * We have `VaListImpl::with_copy` which I can't classify, although it looks like kinda builder method. The struct itself contains pointers, so maybe there is some implied interior mutability, but it's not obvious. 

        * There are also `Path::with_file_name` and `Path::with_extension`, which turn it into `PathBuf`. That's also a kind of builder, but `Path` is _unsized_, which means that builders may only take `&self` or `&mut self` (or some other pointer). Unsized builders are an interesting pattern with basically require `&self`.

        * Finally there is `TestName::with_padding` which basically does a `clone`.

    * tokio --- 4-5 hits, depending on version. The methods are all on types with interior mutability.

    * tracing and tracing-futures --- 1 hit each. Both methods are on a type with interior mutability.

* `&mut self` --- 22 results. Ok, this is getting tiresome. I'll just list the crates.
    * bytes --- 2 hits.
    * indexmap --- 4 hits.
    * miow --- 1 hit.
    * multipart --- 3 hits.
    * time --- 3 hits.
    * tokio --- 1 hit.
    * tokio-native-tls and tokio-tungstenite --- 2 hits each.
    * web-sys --- 1 hit.

The cases themselves all follow the same pattern: those are builder methods which take a mutable builder. Some of them pass the builder through (like `fn with_foo(&mut self, f: Foo) -> &mut Self`), some of them return a different type. The `&mut -> &mut` pattern in particular looks like a very idiomatic way to declare builders since it allows to chain builder methods while at the same time allowing refactoring the configuration into separate steps or separate functions.

TL; DR: there is no self convention for `with_` methods. All kinds of signatures are in use. The numbers are 79/22/127/422. The `&mut self` convention is the rarest, but it also has a strong rationale for builder pattern.

**Comment by Kobzol** [comment]
Thank you for the investigation! Well, this PR is about aligning the inspection with Clippy and Clippy doesn't have any `with` lint... In that direction we should also remove `get` and `set` lints, which would completely undo https://github.com/intellij-rust/intellij-rust/pull/6842.

Since you already have the infrastructure prepared, could you also please search for `get*` and `set*` methods? To see if these make sense in the current form.

**Comment by afetisov** [comment]
With `get_` the results are the following. I've searched with these regular expressions:

```regex
self              fn\s+get\_[^(]+\(\s*self
&self             fn\s+get\_[^(]+\(\s*&\s*self
&mut self         fn\s+get\_[^(]+\(\s*&\s*mut\s*self
get_mut &mut self fn\s+get[^(]+mut\s*\(\s*&\s*mut\s*self
no self           fn\s+get_[^(]+\s*\([\s\n]*[\w&&[^s]]
```

There were also a few variants which included explicit lifetimes, but those gave a negligible number of results.

* `self` --- 94 hits. 
    * 23 of those results are in crates like `futures`, `tokio` etc. They all have the same pattern `get_pin_mut(self: Pin<&mut Self>, ..)`. This is indeed an important pattern: in many cases the type of self will be some smart pointer with a `#[fundamental]` attribute.
    * 52 results are in `std`, only 6 of them a public. 
        * The 46 private functions are methods like `get_unchecked` and `get_unchecked_mut` on various Ranges. Essentially the type that they "get" from is the second argument to the function, and it mostly follows the convention (although it's often a pointer rather than a reference), while the range itself is consumed. I'd take a guess that the reason is that you can't currently implement methods with a `*const Self` as the receiver.
        * The public functions are `Pin::get_ref`, `Pin::get_mut`, `Pin::get_unchecked_mut`, and the `get_unchecked`, `get_unchecked_mut` methods on slice pointers.
    * 18 methods in crates `tonic`, `standback`, `serde_derive` and `clap`, only 5 of them public. These cases are sporadic, I don't see them following any clear pattern. A few are methods on a Copy type.

* `get_..(&self)` --- 532 results, none of which are `get_..._mut(&self)`. Since there are so many and it fits the null hypothesis about the convention, I won't review them in detail.
* `get_...(&mut self)` --- 424 results, only 226 of which follow the expected convention `get_..._mut(&mut self)`. Again, I won't follow on the conforming results. The 198 non-conforming ones have the following properties.
    * 100 hits are in `bytes`, counted twice (2 versions). It has many methods of the form `get_u8(&mut self) -> u8` or `get_int(&mut self, n: usize) -> i64` which read from the buffer, advancing the internal pointer.
    * 24 hits are in winapi. The naming there is all over the place due to backwards compatibility with Win32, so this crate should just be ignored.
    * A notable case is `get_or_insert(_with)` method on HashMap, which occurs both in `hashbrown` and in `std`. The case itself is counted several times, but it looks like a legitimate pattern: how do we name a method which returns a value if present or substitutes a default? Looks like many such cases were avoided by the Entry api  on Maps.
    * There is a similar value `get_key_value` which for some types takes a `&mut self`. Probably it can also do default value insertion.
    * There are a few more hits like `get_flags`, `get_next` or `get_ref` in various crates. Overall I don't see any strong pattern, nor a good reason why those functions couldn't have a different name.
    * Most likely I have counted fewer conforming examples than actually exist, since my regex for this case doesn't allow generic parameters.
    * In general the `get_..._mut(&mut self)` convention looks like the strongest one. I didn't find any cases where a `get_.._mut` would take a `&self`, and cases like `get_..(&mut self)` look rare and not too convincing. The only important exception is `get_or_insert`.
* Finally, `get_` functions which _don't take self at all_. A whopping 1120 cases! Ohhh wee!
    * There are likely some cases not included in this search, since my regex forbids a first argument name starting with "s". I don't expect this to change much.
    * ...but many of these cases are free functions! Frankly, I have no good idea how to filter out the methods.
    * Okay, there is one thing we may try. Rust code is usually well-formatted and follows a common style, which means that methods will be indented at least once.
    * So let's search only for functions which are prefixed by 4 spaces. And may include visibility/safety/async modifiers.
    * This gets us down to only 971 results, many of which are still free functions (inside nested modules or extern blocks). The regex is (mind the prefix spaces)
    
```regex
    [\w\s]*fn\s+get_[^(]+\s*\([\s\n]*[\w&&[^s]]
```

    * 465 results are in `winapi`, which doesn't follow naming conventions. 506 results to go.
    * 380 results are in `web-sys`. It is my understanding that it's also a bindings crate which follows the conventions from the original library, so it can also be ignored. This leaves us with 126 cases.
    * An interesting case in `std` is `Rc::get_mut`. It doesn't have a receiver for standard smart pointer reasons, which makes it signature `get_mut(this: &mut Self) -> &mut T`. If the convention lint stays, this case should be explicitly supported.
    * Careful looking at the other results shows that _almost all_ of them are free functions, external bindings or inner functions (within other functions). Only 22 results which looks like they violate the  convention.

# Resume

It looks like the getter convention is mostly observed, with some notable exception. 

* The most important one is that there can be other receiver types besides `self`, `&self` and `&mut self`, and those types are quite common. If `#[feature(arbitrary_receiver_types)]` is ever stabilized, then there will likely be many more variants. In particular, one can expect `Pin`, `Rc`, `Arc`, `*const` and `*mut` as receivers.
* Another exception is methods without receivers, where the first argument has some `Self`-dependent type. It  is related to the case above.
* `get(self)` vs `get(&self)` doesn't look clear cut. With `Copy` types the choices can be quite varied.
* Functions which accept `&mut self` must have a `mut` suffix, and vice versa. An important exception to this rule is `get_or_insert`.

Phew. That was a mouthful. I think I'll investigate `set` on some other phase of the moon.

**Comment by Kobzol** [comment]
Thanks for the very detailed investigation! Based on these results I'm inclined to remove the `get/set` lints. Let's not create arbitrary heuristics and just closely follow what `Clippy` does, since it's a standard tool used in the Rust ecosystem.

@mchernyavsky WDYT?

**Comment by afetisov** [comment]
One thing to remember is that Rust doesn't follow the getter/setter conventions of Java. Methods which start with `get_` are not getters but rather just some functions. The convention for getters is specified in [Rust API guidelines](https://rust-lang.github.io/api-guidelines/naming.html#getter-names-follow-rust-convention-c-getter). Getters do not use the "get" prefix, they just use the name of the corresponding value. Setters are usually not required, instead you would make a `field_mut` method that returns a mutable reference.

Since methods which start with get/set aren't getters or setters, it is harder to justify the conventions in the lint.

**Review by mchernyavsky — APPROVED** [review]
@Kobzol Thanks! I agree that we should follow `Clippy` conventions.

bors r+

**Comment by jendrikw** [comment]
There is still the warning "methods called `is_*` usually take self by reference or no self; consider choosing a less ambiguous name", where clippy doesn't warn and which is also not mentioned in https://rust-lang.github.io/api-guidelines/naming.html#ad-hoc-conversions-follow-as_-to_-into_-conventions-c-conv. Can you elaborate why you think this warning use useful?

**Comment by Kobzol** [comment]
That corresponds to `is_` here: https://rust-lang.github.io/rust-clippy/master/index.html#wrong_self_convention. Could you post an example code where Clippy doesn't warn and this plugin does? Thanks.

**Comment by jendrikw** [comment]
```rust
trait T {
    fn is_foo(&mut self) -> bool {
        todo!()
    }
}
```

Now that i think about it, my preference would be that clippy warns here.

image

**Comment by Kobzol** [comment]
I also think that `is_` using `&mut self` is a bit suspicious. Maybe it would be better if these warnings could be turned off on a more granular basis (rather than just turning off the whole inspection).

**Comment by jendrikw** [comment]
I think this is because clippy doesn't apply the lint to public items. After setting `avoid-breaking-exported-api = false`, clippy warns for `self`, but not `&mut self`.

**Comment by jendrikw** [comment]
I filed a bug report at rust-lang/rust-clippy#8142.
